src/main/java/com/uid2/admin/auth/AdminAuthMiddleware.java

```diff
@@ -0,0 +1,133 @@
+                    return null;
+                } else {
+                    String givenPrefix = v.substring(0, "bearer ".length());
+                    return !"bearer ".equalsIgnoreCase(givenPrefix) ? null : v.substring("bearer ".length());
+                }
+            }
+        }
+        private boolean isAuthorizedUser(List<String> userGroups) {
+            // TODO temporary, introduce role mapping later
+            return oktaRoles.stream().anyMatch(userGroups::contains);
+        }
+        private boolean isAuthorizedService(List<String> scopes) {
+            if (scopes.contains("uid2.admin.ss-portal") && allowedRoles.contains(Role.SHARING_PORTAL)) {
+                return true;
+            } else if (scopes.contains("uid2.admin.secret-rotation") && allowedRoles.contains(Role.SECRET_MANAGER)) {
+                return true;
+            } else return scopes.contains("uid2.admin.site-sync") && allowedRoles.contains(Role.SECRET_MANAGER);
+        }
+        public void handle(RoutingContext rc) {
+            // human user
+            String idToken = null;
+            if(rc.user() != null && rc.user().principal() != null) {
+                idToken = rc.user().principal().getString("id_token");
+            }
+            if(idToken != null) {
+                validateIdToken(rc, idToken);
+                return;
+            }
+
+            // machine user
+            String authHeaderValue = rc.request().getHeader("Authorization");
+            String accessToken = extractBearerToken(authHeaderValue);
+            if(accessToken == null) {
+                rc.response().setStatusCode(401).end();
+                return;
+            }
+            validateAccessToken(rc, accessToken);
+        }
+
+        private void validateAccessToken(RoutingContext rc, String accessToken) {
+            Jwt jwt;
+            try {
+                jwt = authProvider.getAccessTokenVerifier().decode(accessToken);
+            } catch (JwtVerificationException e) {
+                rc.response().setStatusCode(401).end();
+                return;
+            }
+            if(jwt.getClaims().get("environment") == null || !jwt.getClaims().get("environment").toString().equals(environment)) {
+                rc.response().setStatusCode(401).end();
+                return;
+            }
+            List<String> scopes = (List<String>) jwt.getClaims().get("scp");
+            if(isAuthorizedService(scopes)) {
+                innerHandler.handle(rc);
+            } else {
+                rc.response().setStatusCode(401).end();
+            }
+        }
+
+        private void validateIdToken(RoutingContext rc, String idToken) {
+            Jwt jwt;
+            try {
+                jwt = authProvider.getIdTokenVerifier().decode(idToken, null);
+            } catch (JwtVerificationException e) {
+                rc.session().destroy();
+                rc.response().putHeader("REQUIRES_AUTH", "1").setStatusCode(401).end();
```

Comment:
It is because, for machines using the client_credentials workflow, we always begin the authentication process from the start (no refresh_tokens). For the UI however, we use this to indicate that reauthentication is needed via /login, in the event the refresh_token is expired.